codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,319 @@
+                    if (state.isActiveCountForTree != 0) {
+                        State pState = state(parent);
+                        pState.offerAndInitializePseudoTime(state);
+                        pState.isActiveCountChangeForTree(state.isActiveCountForTree);
+                    }
+                }
+            }
+
+            @Override
+            public void onPriorityTreeParentChanging(Http2Stream stream, Http2Stream newParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    State state = state(stream);
+                    if (state.isActiveCountForTree != 0) {
+                        State pState = state(parent);
+                        pState.remove(state);
+                        pState.isActiveCountChangeForTree(-state.isActiveCountForTree);
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState state) {
+        state(state.stream()).updateStreamableBytes(state.streamableBytes(),
+                                                    state.hasFrame() && state.isWriteAllowed());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        while (connectionState.isActiveCountForTree != 0 && maxBytes > 0) {
+            // connectionState will never have a frame to send, so go right to children.
+            maxBytes -= distributeToChildren(maxBytes, writer, connectionState);
+        }
+
+        return connectionState.isActiveCountForTree != 0;
+    }
+
+    /**
+     * Sets the amount of bytes that will be allocated to each stream. Defaults to 1KiB.
+     * @param minAllocationChunk the amount of bytes that will be allocated to each stream. Must be > 0.
+     */
+    public void allocationQuantum(int allocationQuantum) {
+        if (allocationQuantum <= 0) {
+            throw new IllegalArgumentException("allocationQuantum must be > 0");
+        }
+        this.allocationQuantum = allocationQuantum;
+    }
+
+    private int distribute(int maxBytes, Writer writer, State state, State parentState) throws Http2Exception {
+        if (state.isActive) {
+            int nsent = min(maxBytes, state.streamableBytes);
+            state.write(nsent, writer);
+            if (nsent == 0) {
+                // If a stream sends zero bytes, then we gave it a chance to write empty frames and it is now
+                // considered inactive until the next call to updateStreamableBytes. This allows descendant streams to
+                // be allocated bytes when the parent stream can't utilize them. This may be as a result of the
+                // stream's flow control window being 0.
+                state.updateStreamableBytes(state.streamableBytes, false);
+                return 0;
+            }
+            parentState.pseudoTime += nsent;
```

Comment:
Updating `pseudoTime` here doesn't feel right, because it isn't involved with recursion. When a grandchild writes, `nsent` should increment all the `pseudoTimes` up the tree. It looks like maybe the proper place for this line is just before the call to `updatePseudoTime()`, which would make sense to update the parent at the same time as the child.

That seems like it will work, even for the connection, but only because the connection never `isActive == true`. A bit subtle, but it also makes sense...